### PR TITLE
[FIX] onboarding: prevent crash with multiple progress records

### PR DESCRIPTION
**Pull request by bda-odoo**
Multiple progress records for the same onboarding and company_id (False) were allowed to be created in several databases which resulted in crashes.

As simple stable fix, we'll select only the latest onboarding_progress records created per onboarding when trying to compute the current_progress_id and similarly for progress steps.

The [fix](https://github.com/odoo/odoo/pull/108183) in this addresses the fix up to version forward bot 16.1. However, in my case starting the database from versions 16.4 and 17.0, this code is no longer present, as confirmed by TBG. As a result, the same issue reappears, which is why I prepared this fix.

tbg-1565
upg-2106731,2094548

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by flch-odoo** [comment]
Would you mind adding a reference to the original fix you've clearly got a lot of inspiration from?

In addition, perhaps a note explaining why this wasn't completely fixed with these earlier fixes in community and upgrade? Thanks.

**Comment by bda-odoo** [comment]
> Would you mind adding a reference to the original fix you've clearly got a lot of inspiration from?
> 
> In addition, perhaps a note explaining why this wasn't completely fixed with these earlier fixes in community and upgrade? Thanks.

Thank you for the review @flch-odoo . The [fix](https://github.com/odoo/odoo/pull/108183) in this addresses the fix up to version forward bot 16.1. However, in my case starting the database from versions 16.4 and 17.0, this code is no longer present, as confirmed by TBG. As a result, the same issue reappears, which is why I prepared this fix.

**Comment by flch-odoo** [comment]
1. For the original fix, I was thinking of an in-commit message reference, please.
2. This code was cleaned because we had added a unicity constraint in the `init` of `onboarding.progress`, and `onboarding.progress.step`, already present in 16.4. So, do we know exactly how this happened again?

**Comment by bda-odoo** [comment]
> 1. For the original fix, I was thinking of an in-commit message reference, please.
> 2. This code was cleaned because we had added a unicity constraint in the `init` of `onboarding.progress`, and `onboarding.progress.step`, already present in 16.4. So, do we know exactly how this happened again?

In the demo database, I can't reproduce because constraints violent not a possible way to create multiple progress. In the customer database, we have two requests which I mention PR description to be honest I don't understand how customers create those faulty records if the standard constraints are still in place.And also investigation customer production database this error still there.Please correct me If I'm wrong.

**Comment by flch-odoo** [comment]
From the requests, I saw that the problem is that one record has set company and the other doesn't, which is why the unicity constraint didn't prevent this from happening, and I can't figure out simply how this happened.

This means however, that the fix should instead use the record with set `company_id` or not depending on the onboarding (step) being `is_per_company`, and NOT a create-time based approach.

In other words,
If there is any step of an onboarding with `is_per_company==True`, the related `onboarding.progress` and `onboarding.progress.step` to use are those with `company_id` and the other otherwise.

**Comment by bda-odoo** [comment]
> From the requests, I saw that the problem is that one record has set company and the other doesn't, which is why the unicity constraint didn't prevent this from happening, and I can't figure out simply how this happened.
> 
> This means however, that the fix should instead use the record with set `company_id` or not depending on an onboarding (step) being `is_per_company`, and NOT a create-time based approach.
> 
> In other words, If there is any step of an onboarding with `is_per_company==True`, the related `onboarding.progress` and `onboarding.progress.step` to use are those with `company_id` and the other otherwise.

@flch-odoo Thank you for your valuable input. For now, I am closing this PR. This will be considered as a specific case; if there are more requests on the platform in the future, we will look into it further.
